docs/book/src/user/quick-start.md

```diff
@@ -285,10 +285,13 @@ The Docker provider is not designed for production use and is intended for devel
 
 </aside>
 
-The Docker provider does not require additional prerequisites.
-You can run:
+The Docker provider does requires to enable the `ClusterTopology` feature to support cluster upgrades from kubernetes <= v1.23 to v1.24.
```

Comment:
why does Docker provider require enabling an experimental feature to upgrade to the latest k8s version? Will this apply to other providers as well or is this just for Docker? What about MachinePools (ClusterClass doesn't currently support MachinePool)?

Reply:
This only applies to the docker provider / CAPD.

The TLDR is: currently we hardcode `cgroupDriver: cgroupfs` in all templates for CAPD but this does not work anymore for kubernetes >= v1.24 because of a breaking change in kind https://github.com/kubernetes-sigs/kind/releases/tag/v0.13.0 :

> * systemd cgroups driver will be used for Kubernetes v1.24.0+ (rather than 1.21.0+ when https://github.com/kubernetes/kubernetes/pull/99471, which we previously overrode).
>   * NOTE: You must use kind v0.13.0+ with Kubernetes v1.24.0+ images, and if you built your own Kubernetes v1.24.0+ image with a previous kind version you will need to re-built when switching to kind v0.13.0+.
>   * NOTE: You do not need to be using systemd on the host machine, systemd is used inside the kind node containers. We are now using it for Kubernetes pods in addition to running kubelet, containerd etc.
>   * KIND will continue to use cgroupfs for Kubernetes versions prior to v1.24.0.

Some history:

* kubeadm started to default to `cgroupDriver: systemd` for Kubernetes >= v1.21 (which is what the other providers also use afaik)
* Kind and its images which are used for CAPD did not support `cgroupDriver: systemd` for Kubernetes <= v1.23
* Because of that all templates for CAPD did hardcode `cgroupDriver: cgroupfs`, but beginning with v1.24 this does not work anymore.
* Therefor Kubernetes >= v1.24 requires us to not set `cgroupDriver: cgroupfs` anymore.

Perspective for the Docker/CAPD quickstart guide:

* If somebody uses the quickstart guide and provisions a cluster having version v1.23
* To upgrade the cluster to v1.24 the person would have to:
  * Remove the `cgroupDriver: cgroupfs` entries at the KCP
  * Change the kubernetes version at the KCP
  * Create a new `KubeadmConfigTemplate` for the `MachineDeployment`
  * Change the `MachineDeployment` to refer the new `KubeadmConfigtemplate`
  * Change the kubernetes version at the `MachineDeployment`

With ClusterClass and the version-aware patch for `cgroupDriver`: a user would only need to change the kubernetes version at the Cluster object.

Regarding MachinePools in Docker/CAPD:

* When a `MachinePool` gets updates from v1.23 -> v1.24: the same steps as above for `MachineDeployment` are required (new `KubeadmConfigTemplate` and changing `MachinePool`)
* As soon as #5991 is implemented, an version-aware patch could also get added for `MachinePools`

Perspective for the E2E tests:

* Its way easier and requires less temporary hacks (until Kubernetes <= v1.23 does not get tested anymore) to use ClusterClass and the version aware patch
* Only CAPD tests using MachinePools require the two temporary hacks which are in this PR.

I hope this brings some light in the dark and helps to continue the discussion.

Some more context is at #6509 :-)

If we decide to continue as is I could of course add some more information at the quickstart guide about why it is required to enable ClusterTopology for docker/CAPD.

Reply:
Created Slack thread for it: https://kubernetes.slack.com/archives/C8TSNPY4T/p1653416265784229

Reply:
Based on the thread I only kept the ClusterClass template. Please let me know if there are any objections (@CecileRobertMichon @fabriziopandini)